src/v/rpc/transport.cc

```diff
@@ -192,33 +193,41 @@ transport::do_send(sequence_t seq, netbuf b, rpc::client_opts opts) {
       });
 }
 
-void transport::dispatch_send() {
+void transport::dispatch_send(const client_opts& opts) {
     ssx::background
-      = ssx::spawn_with_gate_then(_dispatch_gate, [this]() mutable {
-            return ss::do_until(
-              [this] {
-                  return _requests_queue.empty()
-                         || _requests_queue.begin()->first
-                              > (_last_seq + sequence_t(1));
-              },
-              [this] {
-                  auto it = _requests_queue.begin();
-                  _last_seq = it->first;
-                  auto buffer = std::move(it->second->buffer).get();
-                  auto units = std::move(it->second->resource_units);
-                  auto v = std::move(*buffer).as_scattered();
-                  auto msg_size = v.size();
-                  _requests_queue.erase(it->first);
-                  return _out.write(std::move(v))
-                    .finally([this, msg_size, units = std::move(units)] {
-                        _probe.add_bytes_sent(msg_size);
-                    });
-              });
-        }).handle_exception([this](std::exception_ptr e) {
-            vlog(rpclog.info, "Error dispatching socket write:{}", e);
-            _probe.request_error();
-            fail_outstanding_futures();
-        });
+      = ssx::spawn_with_gate_then(
+          _dispatch_gate,
+          [this, t = opts.timeout]() mutable {
+              return ss::do_until(
+                [this] {
+                    return _requests_queue.empty()
+                           || _requests_queue.begin()->first
+                                > (_last_seq + sequence_t(1));
+                },
+                [this, t = std::move(t)] {
+                    auto it = _requests_queue.begin();
+                    _last_seq = it->first;
+                    auto buffer = std::move(it->second->buffer).get();
+                    auto units = std::move(it->second->resource_units);
+                    auto v = std::move(*buffer).as_scattered();
+                    auto msg_size = v.size();
+                    _requests_queue.erase(it->first);
+                    auto f = _out.write(std::move(v));
+                    return ss::with_timeout(t, std::move(f))
+                      .handle_exception([](const std::exception_ptr& e) {
```

Comment:
I think that here we should use `handle_exception_type` and handle only timeout exceptions, now each timeout will cause all the outstanding requests to fail.

Reply:
you are right